gadopt/scalar_equation.py

```diff
@@ -93,165 +72,96 @@ class ScalarDiffusionTerm(BaseTerm):
-
-        return -F
-
-
-class ScalarSourceTerm(BaseTerm):
+    kappa = eq.diffusivity
+    dim = eq.mesh.geometric_dimension()
+    diff_tensor = kappa if len(kappa.ufl_shape) == 2 else kappa * Identity(dim)
+
+    reference_for_diffusion = getattr(eq, "reference_for_diffusion", 0)
+    q = trial + reference_for_diffusion
+
+    F = inner(grad(eq.test), dot(diff_tensor, grad(q))) * eq.dx
+
+    sigma = interior_penalty_factor(eq, shift=-1)
+    if not is_continuous(eq.trial_space):
+        sigma_int = sigma * avg(FacetArea(eq.mesh) / CellVolume(eq.mesh))
+        F += (
+            sigma_int
+            * inner(jump(eq.test, eq.n), dot(avg(diff_tensor), jump(q, eq.n)))
+            * eq.dS
+        )
+        F -= inner(avg(dot(diff_tensor, grad(eq.test))), jump(q, eq.n)) * eq.dS
+        F -= inner(jump(eq.test, eq.n), avg(dot(diff_tensor, grad(q)))) * eq.dS
+
+    for bc_id, bc in eq.bcs.items():
+        if "q" in bc and "flux" in bc:
+            raise ValueError("Cannot apply both `q` and `flux` on the same boundary.")
+
+        if "q" in bc:
+            jump_q = trial - bc["q"]
+            sigma_ext = sigma * FacetArea(eq.mesh) / CellVolume(eq.mesh)
+            # Terms below are similar to the above terms for the DG dS integrals.
+            F += (
+                2
+                * sigma_ext
+                * eq.test
+                * inner(eq.n, dot(diff_tensor, eq.n))
+                * jump_q
+                * eq.ds(bc_id)
+            )
+            F -= inner(dot(diff_tensor, grad(eq.test)), eq.n) * jump_q * eq.ds(bc_id)
+            F -= inner(eq.test * eq.n, dot(diff_tensor, grad(q))) * eq.ds(bc_id)
+        elif "flux" in bc:
+            # Here we need only the third term because we assume jump_q = 0
+            # (q_ext = trial) and flux = kappa dq/dn = dot(n, dot(diff_tensor, grad(q)).
+            F -= eq.test * bc["flux"] * eq.ds(bc_id)
+
+    return -F
+
+
+def source_term(eq: Equation, trial: Argument | ufl.indexed.Indexed | Function) -> Form:
     r"""Scalar source term `s_T`."""
+    F = -dot(eq.test, eq.source) * eq.dx
 
-    def residual(
-        self,
-        test: fd.ufl_expr.Argument,
-        trial: fd.ufl_expr.Argument | fd.Function,
-        trial_lagged: Optional[fd.ufl_expr.Argument | fd.Function] = None,
-        fields: Optional[dict[str, fd.Constant | fd.Function]] = None,
-        bcs: Optional[dict[int, dict[str, int | float]]] = None,
-    ) -> fd.ufl.core.expr.Expr:
-        if 'source' not in fields:
-            return 0
-        phi = test
-        source = fields['source']
+    return -F
 
-        # NOTE, here source term F is already on the RHS
-        F = dot(phi, source)*self.dx
 
-        return F
+def absorption_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
+    r"""Scalar absorption term `\alpha_T T`."""
+    # Implement absorption term implicitly at current time step.
+    F = dot(eq.test, eq.sink * trial) * eq.dx
```

Comment:
Sounds fair to me. What about `sink_coeff`? Or would you prefer `absorption_coeff`?

Reply:
actually on second thoughts it's probably a bad idea to change the user interface. You've managed so far to keep everything the same from the demo perspective which is a good thing. I guess if it was raised at something like the gadopt workshop or multiple mention that it is weird we should change it in a separate PR

Reply:
I am not sure I follow. In `master`, this term is currently called `absorption_coefficient` (see [here](https://github.com/g-adopt/g-adopt/blob/77d9d524f2547b0d167cc17dbbff804d2e06c58e/gadopt/energy_solver.py#L93) and [here](https://github.com/g-adopt/g-adopt/blob/77d9d524f2547b0d167cc17dbbff804d2e06c58e/gadopt/scalar_equation.py#L207)).

Reply:
I just mean that the keyword used in approximation is 'sink' so it proably makes most sense to be consistent with that... otherwise sink_coeff is good

Reply:
Well, we have `linearized_energy_sink` in `approximations.py`, but we have `absorption_term` in `scalar_equation`. Should that be homogenised too?

Reply:
I guess sink_term is easier to write than absorption_term!